include/ginkgo/core/base/math.hpp

```diff
@@ -444,11 +451,17 @@ GKO_INLINE __device__ constexpr T zero(const T &)
  * @return the multiplicative identity for T
  */
 template <typename T>
-GKO_INLINE __device__ constexpr T one()
+GKO_INLINE __device__ constexpr xstd::enable_if_t<!is_complex<T>(), T> one()
 {
     return T(1);
 }
 
+template <typename T>
+GKO_INLINE __device__ constexpr xstd::enable_if_t<is_complex<T>(), T> one()
+{
+    return T(one<remove_complex<T>>());
+}
```

Comment:
```
complexT(0); // failed
T zero = 0;
complexT(zero); // ok
```
is it correct?

Reply:
The issue is more complex: For some reason, HCC prefers the `__device__ zero` over the `__host__ zero` when calling it from a `__host__` context, which fails as only the implicit constructors of `std::complex` are visible then. I'm still investigating possible solutions, a simple `enable_if<!is_same<std::complex<...>,T>` causes some other issues.